launch/launch/actions/execute_process.py

```diff
@@ -382,7 +384,21 @@ def _shutdown_process(self, context, *, send_sigint):
         if self.__shutdown_received:
             # Do not handle shutdown more than once.
             return None
+
+        if self.__running and (self.process_details is None or self._subprocess_transport is None):
+            # Defer shut down if the process is not ready
+            context.register_event_handler(
+                OnProcessStart(
+                    on_start=lambda event, context:
+                    self._shutdown_process(context, send_sigint=send_sigint)))
+            return None
+
         self.__shutdown_received = True
+
+        if not self.__running:
+            # Not running, skip shutdown
+            return None
```

Comment:
The logic can be more clear if it's reordered:

```python3
        if self.__shutdown_received:
            # Do not handle shutdown more than once.
            return None
        self.__shutdown_received = True
        if not self.__running:
            return
        if self.process_details is None or self._subprocess_transport is None:
            # Defer shut down if the process is not ready
            context.register_event_handler(
                OnProcessStart(
                    on_start=lambda event, context:
                    self._shutdown_process(context, send_sigint=send_sigint)))
            return None
        ...
```

Reply:
That was my original intention, but the block has to be called before the `self.__shutdown_received = True` to be able of deferring the shutdown process.